designs/node-termination-hook.md

```diff
@@ -0,0 +1,72 @@
+# Custom Node termination (custom finalizer)
+
+## Motivation
```

Comment:
I'm curious why we think that this interface needs to be "pluggable" rather than supporting these features natively in Karpenter directly. I know we have linked out a few features here, but I'd be interested to hear the tangible ways that controllers are going to plug-in to this new system, so that we know that we are actually solving these use-cases if we were to implement this interface. IMO, a lot of these features make sense to be part of Karpenter natively, we just haven't had the time or resources to get some of them into the project. There's also some design thinking that needs to go into things like "cordon nodes after expiry."

I think there's two buckets of features that are being listed here: 1) Features that we know that we want to do, but need a little design and direction to determine the exact best way to move forward with it (https://github.com/kubernetes-sigs/karpenter/issues/743, https://github.com/kubernetes-sigs/karpenter/issues/622) and 2) Features that we're less sure that we want to pursue, with design that might touch both upstream and Karpenter (https://github.com/kubernetes-sigs/karpenter/issues/740, https://github.com/kubernetes-sigs/karpenter/issues/690, https://github.com/aws/karpenter-provider-aws/issues/5232).

The other thing to mention, is that any controller today has the ability to plug an additional finalizer into the Node or the NodeClaim to add additional finalization logic. Granted, this won't stop the Karpenter finalizer from taking action, but you can layer additional things on top of the typical finalization flow.

Reply:
_TLDR: Can we modularize finalizer architecture to be more Singular Responsibility so contribution may be less complex/ less risky and users can attempt to take over parts they want to customize?_

Good question, sorry I never answer similar ask from @njtran. 
Initially , we were thinking about a way to notify operators that a node is stuck in Production rather than relying on a timeout to "do something" with the node that's stuck. In theory we could update Karpenter's finalizer to send a more clear event and we can capture that to notify our operators.  However this begs the question if there are ways we can have better control drain and node termination. 

Specifically, I understand the desire for Karpenter to natively do more advance termination logic however it is difficult to do this right now because the termination logic is tightly wrapped into a single finalizer. I wonder if alternative approach could be breaking out the finalizer into "[stages](https://github.com/kubernetes-sigs/karpenter/pull/843/files#r1421083940)" and allowing contributors to contribute logic that can perform these stages differently?

Reply:
Thinking about this more, I suspect my thinking is bias towards how we did things pre-Karpenter. Since we use ASGs, we rely heavily on lifecycle hooks which in a way I am trying to replace with finalizer. I think finalizer isn't completely the replacement and maybe it shouldn't be. 

I think we could move a majority of the logic to be on a daemonset that interacts with host level processes since finalizer does wait for daemonsets to terminate. But this raise a question around [graceful termination timeout](https://github.com/kubernetes-sigs/karpenter/pull/834). I see that the finalizer [orders](https://github.com/kubernetes-sigs/karpenter/blob/main/pkg/controllers/node/termination/terminator/terminator.go#L113) it termination sequence. If we want do move in that direction (termination lifecycle hook logic -> daemonset) we need our daemonset to gracefully terminate which means the graceful termination period will likely cause issues if user pods drain the full termination period. I think we (my team) may want to separate out finalizer between user application vs cluster operator applications graceful termination periods.

Does it make sense for finalizer to capture all these complexity? I am just trying to wrap my head around how to keep Karpenter's API surface clean but yet not be overly complex in what it manages because of user needs.

Reply:
I think this RFC is prescribing a solution rather than tackling the problem as a whole, which I understand to be a custom drain flow. I think a custom finalizer could be one of the solutions here, but I think we're leaving out some considerations of other alternatives and an understanding of the use-cases here. 

Here are some questions I find myself asking, which I'm happy to talk over slack if that makes the discussion easier:
- What are the alternatives for API Spec here? 
  - A setting for enabling/disabling the termination controller response to finalizers? 
  - A new finalizer to indicate a custom drain flow? 
  - A setting to say which part of each termination action needs to be different?
- Do all users who want custom termination flow have the ability to run a custom controller that handles the termination flow themselves? Is full extensibility the right answer? 
  - If users want all existing features of Karpenter, but a change in one part of the code, does a fully custom termination flow solve this issue for them, or would that be not worth the effort for them?
- What termination controller features do _you_ need implemented in Karpenter? Can we do any of those without adding an entirely new controller? Are any of those use-cases not covered in an issue?